amr-wind/wind_energy/actuator/disk/ActSrcDiskOp.H

```diff
@@ -0,0 +1,133 @@
+
+public:
+    explicit ActSrcOp(typename ActTrait::DataType& data)
+        : m_data(data)
+        , m_act_src(m_data.sim().repo().get_field("actuator_src_term"))
+    {}
+
+    void initialize();
+
+    void setup_op() { copy_to_device(); }
+
+    void operator()(
+        const int lev, const amrex::MFIter& mfi, const amrex::Geometry& geom);
+};
+
+template <typename ActTrait>
+void ActSrcOp<
+    ActTrait,
+    ActSrcDisk,
+    typename std::enable_if<std::is_base_of<DiskType, ActTrait>::value>::type>::
+    initialize()
+{
+    const auto& grid = m_data.grid();
+    m_pos.resize(grid.pos.size());
+    m_force.resize(grid.force.size());
+}
+
+template <typename ActTrait>
+void ActSrcOp<
+    ActTrait,
+    ActSrcDisk,
+    typename std::enable_if<std::is_base_of<DiskType, ActTrait>::value>::type>::
+    copy_to_device()
+{
+    const auto& grid = m_data.grid();
+
+    amrex::Gpu::copy(
+        amrex::Gpu::hostToDevice, grid.pos.begin(), grid.pos.end(),
+        m_pos.begin());
+    amrex::Gpu::copy(
+        amrex::Gpu::hostToDevice, grid.force.begin(), grid.force.end(),
+        m_force.begin());
+}
+
+template <typename ActTrait>
+void ActSrcOp<
+    ActTrait,
+    ActSrcDisk,
+    typename std::enable_if<std::is_base_of<DiskType, ActTrait>::value>::type>::
+operator()(const int lev, const amrex::MFIter& mfi, const amrex::Geometry& geom)
+{
+    const std::string fname = ActTrait::identifier() + ActSrcDisk::identifier();
+    BL_PROFILE("amr-wind::ActSrcOp<" + fname + ">");
+
+    const auto& bx = mfi.tilebox();
+    const auto& sarr = m_act_src(lev).array(mfi);
+    const auto& problo = geom.ProbLoArray();
+    const auto& dx = geom.CellSizeArray();
+
+    auto& data = m_data.meta();
+
+    const amrex::Real dR = data.dr;
+    const amrex::Real epsilon = data.epsilon;
+    const vs::Vector m_normal(data.normal_vec);
+    const vs::Vector m_origin(data.center);
+    const auto* pos = m_pos.data();
+    const auto* force = m_force.data();
+    const int npts = data.num_force_pts;
+
+    amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+        const vs::Vector cc{
+            problo[0] + (i + 0.5) * dx[0],
+            problo[1] + (j + 0.5) * dx[1],
+            problo[2] + (k + 0.5) * dx[2],
+        };
+
+        amrex::Real src_force[AMREX_SPACEDIM]{0.0, 0.0, 0.0};
+        for (int ip = 0; ip < npts; ++ip) {
+            const auto R =
+                utils::delta_pnts_cyl(m_origin, m_normal, m_origin, pos[ip])
+                    .x();
+            const auto dist_on_disk =
+                utils::delta_pnts_cyl(m_origin, m_normal, cc, pos[ip]);
+            const auto& pforce = force[ip];
+
+            const amrex::Real weight_R =
+                utils::linear_basis_1d(dist_on_disk.x(), dR);
+            const amrex::Real weight_T =
+                1.0 / (::amr_wind::utils::two_pi() * R);
+            const amrex::Real weight_N =
+                utils::gaussian1d(dist_on_disk.z(), epsilon);
```

Comment:
It just occurred to be that there's a possbility for the gaussian in the z direction could interfere with the upstream blockage measurements, especially if the gaussian has a long tail.  This is something of a hypothetical at this point, and we'll probably need to study the effect of the epsilon.  But we may need to use compact functions, or gaussians with cut-off tails if that is a problem.

Reply:
@lawrenceccheung yes that is a good point. My goal is to make this modular enough to select the spreading function in each principle direction in the input deck. So we could add a more compact function if we'd like. Really with this Gaussian it just needs to be selected to match the mesh requirements so it can still be relatively compact.